[LLD] [COFF] Add tests to observe details about LTO and __imp_ prefixes. NFC.

mstorsjo opened the pull request:
The commit 3ab6209a3f93bdbeec8e9b9fcc00a9a4980915ff had the undesired
effect of retaining every `__imp_` symbol, even if it isn't referenced
in any way. Add a testcase to observe this behaviour, to serve as
a reference point if this behaviour were to be improved later.

Port the testcase from b963c0b658cc54b370832df4f5a3d63fd69da334 from
the llvm/LTO testsuite into LLD as a preparation for changing that fix;
the moved testcase has a comment for one case which doesn't work
currently.

The testcase is ported mostly as is, but with symbol mangling simplified,
rewriting function names from MSVC C++ mangling to plain C, and
unnecessary debug info is removed.

Add a case of a dllimported data symbol, in addition to the existing
call of a dllimported function.

Also extend the testcase to test combinations of both regular
object files and LTO objects. Leave out one combination which currently
fails, with a comment.

MaskRay commented:
You can generalize the patch "[LLD] [COFF] Add a test to observe an unwanted effect of 3ab6209. NFC" to include `lld/test/COFF/lto-dllimport.ll`  as well, and rebase "[[LTO] [LLD] Don't alias the __imp_func and func symbol resolutions" to show the update to `lto-dllimport.ll`

mstorsjo committed:
[LLD] [COFF] Add tests to observe details about LTO and __imp_ prefixes. NFC.

The commit 3ab6209a3f93bdbeec8e9b9fcc00a9a4980915ff had the undesired
effect of retaining every __imp_ symbol, even if it isn't referenced
in any way. Add a testcase to observe this behaviour, to serve as
a reference point if this behaviour were to be improved later.

Port the testcase from b963c0b658cc54b370832df4f5a3d63fd69da334 from
the llvm/LTO testsuite into LLD as a preparation for changing that fix;
the moved testcase has a comment for one case which doesn't work
currently.

The testcase is ported mostly as is, but with symbol mangling simplified,
rewriting function names from MSVC C++ mangling to plain C, and
unnecessary debug info is removed.

Add a case of a dllimported data symbol, in addition to the existing
call of a dllimported function.

Also extend the testcase to test combinations of both regular
object files and LTO objects. Leave out one combination which currently
fails, with a comment.

MaskRay commented:
This lacks a test for an `__imp_` alias to a data symbol (LLVM global variable), right? It seems useful to add one, even if the behavior is the same as a function symbol, as at the object file format level there is some nuance.

mstorsjo commented:
> This lacks a test for an `__imp_` alias to a data symbol (LLVM global variable), right?

Sure, I updated it with such a case, will merge this and #71376 once the CI has passed.

> It seems useful to add one, even if the behavior is the same as a function symbol, as at the object file format level there is some nuance.

Actually, on the object file format level, they are identical. The main gotcha where they differ, is if you reference a symbol without the dllimport attribute, but it turns out to be imported from a DLL. For a function, this just causes the function call to go via a thunk, while for data symbols, it's a linker error. (In MinGW contexts, it can get fixed with the autoimport/runtime pseudo relocation mechanism though.)
